### ISSUE TITLE
transform_to_real_path broken

### ISSUE THREAD

**Issue by alfink** [qilingframework/qiling#446]
There are a bunch of bugs in transform_to_real_path on the dev branch that break all my scripts

# symlinks
This is the code on the dev branch:
```python
        if isinstance(real_path, os.PathLike) and os.path.islink(real_path):
            link_path = os.readlink(real_path)
            if link_path.is_absolute():
                real_path = Path(link_path)
            else:
                real_path = Path(os.path.join(os.path.dirname(path), link_path))
```
- os.readlink returns a string, so link_path.is_absolute() will throw an error. Replacing `link_path.is_absolute()` with `Path(link_path).is_absolute()` works for me
- `real_path = Path(os.path.join(os.path.dirname(path), link_path))` returns the wrong path, as `path` is the user-provided relative path. you want to use `real_path` instead. So the line should be `real_path = Path(os.path.join(os.path.dirname(real_path), link_path))`

# fs mapping

I wanted to do a pull request for the fixes above, but the fs mapping is still broken and I don't want to do the PR before testing if my fixes work. None of my fs_mappings work anymore with the current dev branch. I think this piece of code is not really working
```python
        # TODO: A better design for fs mapping.
        for fm, to in self.ql.fs_mapper:
            if to == path and not isinstance(fm, str):
                real_path = fm
                return real_path
            elif isinstance(fm, str):
                try:
                    remains = relative_path.relative_to(Path(fm))
                    real_path = Path(to) / remains
                except ValueError:
                    continue
```
realtive_path.relative_to fails whenever relative_path is not a subpath of fm, which makes it always fail for me. Also, I'm not sure if the if-branch is correct, because the function wouldn't return a string then.

**Comment by xwings** [comment]
I guess i need to ping @wtdcode , the reason why we update this code is to fix the common path conversion between linux and windows.

**Comment by xwings** [comment]
Ok, you are right. It breaks all the fs_mapper. 

Iwill  need to add into CI test and make sure we dont break that again.

**Comment by wtdcode** [comment]
Thanks for pointing out the two bugs in symlinks. For fs mapping, as comments suggest, I think we should have a better implementation here. I will draft a PR these days.

**Comment by xwings** [comment]
@alfink either way the fs_mapper is broken for now.

Will you be able to submit a PR the two fix u mention and we will work on fixing the fs_map ?

**Comment by xwings** [comment]
i think the fs_mapper works now.

@alfink will you be able to try ?

**Comment by xwings** [comment]
close for  now

**Comment by alfink** [comment]
Its better now, but still broken. Look at this function:
```python
    def _open_mapping_ql_file(self, ql_path, openflags, openmode):
        real_dest = self._mapping[ql_path]
        if not isinstance(real_dest, str): # We have opened this mapping or it is implemented by user.
            return real_dest
        else:
            qlfile = ql_file.open(real_dest, openflags, openmode) # open the path and replace the destination now.
            self._mapping[ql_path] = qlfile
            return qlfile
```

Assume /dev/urandom is mapped. When the program opens it, it will return a new ql_file qlfile and it sets self._mapping["/dev/urandom"] = qlfile. When /dev/urandom is opened again, self._mapping["/dev/urandom"] is still a reference to qlfile, so it won't open it again, but returns a reference to the ql_file of the first open. Now we have the situation that there are 2 file decriptors in the emulated code, but they share the same file descriptor on the host. As soon as any of both emulated file descriptors are closed it will close the shared one on the host, and all following reads, writes, closes, etc. on the still open emulated file descriptor will fail.

**Comment by wtdcode** [comment]
Sounds reasonable indeed. I will consider a better way to implement it.

________________________________

**Comment by alfink** [comment]
Commenting out following writes to _mapping seem to fix it for me:
```python
    def _open_mapping_ql_file(self, ql_path, openflags, openmode):
        real_dest = self._mapping[ql_path]
        if not isinstance(real_dest, str): # We have opened this mapping or it is implemented by user.
            return real_dest
        else:
            qlfile = ql_file.open(real_dest, openflags, openmode) # open the path and replace the destination now.
            #self._mapping[ql_path] = qlfile
            return qlfile
    
    def _open_mapping(self, ql_path, openmode):
        real_dest = self._mapping[ql_path]
        if not isinstance(real_dest, str):
            return real_dest
        else:
            f = open(real_dest, openmode)
            #self._mapping[ql_path] = f
            return f
```

**Comment by wtdcode** [comment]
Of course. Since we won’t remember the paths we have opened if you comment that line. I think we should have another test for opening a mapping multiple times.

Besides, how about user-defined mapped objects for two different file descriptors? I mean, if a path, say ‘/dev/urandom’, has been opened twice, shall we return two distinct instances of the same class, or simply return the same instance? Maybe asking user to pass a class, not an object is the solution?

________________________________

**Comment by alfink** [comment]
I think you're talking of the `QlFsMappedObject` class. My intuition would be to pass a subclass that inherits from `QlFsMappedObject`, which is is then used to create a new instance on every open. Because then its easier to keep for example track of the position in the file, because you can simply create a field for that

But I haven't used that class before. I only use strings for my fs_map currently.

**Comment by xwings** [comment]
@alfink will you be able to double check on your side ?
